qiskit/chemistry/ground_state_calculation/ground_state_calculation.py

```diff
@@ -0,0 +1,99 @@
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""The ground state calculation interface."""
+
+from abc import ABC, abstractmethod
+from typing import Dict, List, Any, Optional, Union
+
+import numpy as np
+
+from qiskit import QuantumCircuit
+from qiskit.circuit import Instruction
+from qiskit.quantum_info import Statevector
+from qiskit.result import Result
+from qiskit.aqua.operators import OperatorBase, WeightedPauliOperator
+from qiskit.chemistry.drivers import BaseDriver
+from qiskit.chemistry.results import EigenstateResult
+
+from ..qubit_transformations.qubit_operator_transformation import QubitOperatorTransformation
+
+
+class GroundStateCalculation(ABC):
+    """The ground state calculation interface"""
+
+    def __init__(self, transformation: QubitOperatorTransformation) -> None:
+        """
+        Args:
+            transformation: transformation from driver to qubit operator (and aux. operators)
+        """
+        self._transformation = transformation
+
+    @property
+    def transformation(self) -> QubitOperatorTransformation:
+        """Returns the transformation used to obtain a qubit operator from the molecule."""
+        return self._transformation
+
+    @transformation.setter
+    def transformation(self, transformation: QubitOperatorTransformation) -> None:
+        """Sets the transformation used to obtain a qubit operator from the molecule."""
+        self._transformation = transformation
+
+    @abstractmethod
+    def compute_groundstate(self, driver: BaseDriver,
+                            aux_operators: Optional[List[Any]] = None
+                            ) -> EigenstateResult:
+        """Compute the ground state energy of the molecule that was supplied via the driver.
+
+        Args:
+            driver: a chemistry driver object which defines the chemical problem that is to be
+                    solved by this calculation.
+            aux_operators: Additional auxiliary operators to evaluate. Must be of type
+                ``FermionicOperator`` if the qubit transformation is fermionic and of type
+                ``BosonicOperator`` it is bosonic.
+
+        Returns:
+            An eigenstate result.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def returns_groundstate(self) -> bool:
+        """Whether this class returns only the ground state energy or also the ground state itself.
+
+        Returns:
+            True, if this class also returns the ground state in the results object.
+            False otherwise.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def evaluate_operators(self,
```

Comment:
Is there a good reason to have this on a ground state interface? It seems like a pretty generic function that would be more suited as a utility somewhere? The state can evidently be any state and need not be limited to a ground state.

Reply:
The operators are evaluated using the same "runtime environment" as the ground state calculation (which may be e.g. a quantum instance or a numpy eigensolver). If we don't put this in the GSC we would have to add and inject a separate `StateEvaluator`  that can either be numpy- or quantum- based. That would rather be a parent class for (min)eigensolvers but not just a stand-alone function. Thus, having the function attached to the GSC seems the most convenient solution.